optuna/visualization/_contour.py

```diff
@@ -303,4 +240,157 @@ def _generate_contour_subplot(
+
+    return _ContourInfo(sorted_params=sorted_params, sub_plot_infos=sub_plot_infos)
+
+
+def _get_contour_subplot_info(
+    trials: List[FrozenTrial],
+    x_param: str,
+    y_param: str,
+    target: Optional[Callable[[FrozenTrial], float]],
+) -> _SubContourInfo:
+
+    xaxis = _get_axis_info(trials, x_param)
+    yaxis = _get_axis_info(trials, y_param)
+
+    if x_param == y_param:
+        return _SubContourInfo(xaxis=xaxis, yaxis=yaxis, z_values={})
+
+    if len(xaxis.indices) < 2:
+        _logger.warning("Param {} unique value length is less than 2.".format(x_param))
+        return _SubContourInfo(xaxis=xaxis, yaxis=yaxis, z_values={})
+    if len(yaxis.indices) < 2:
+        _logger.warning("Param {} unique value length is less than 2.".format(y_param))
+        return _SubContourInfo(xaxis=xaxis, yaxis=yaxis, z_values={})
+
+    z_values = {}
+    for i, trial in enumerate(trials):
+        if x_param not in trial.params or y_param not in trial.params:
+            continue
+        x_value = xaxis.values[i]
+        y_value = yaxis.values[i]
+        assert x_value is not None
+        assert y_value is not None
+        x_i = xaxis.indices.index(x_value)
+        y_i = yaxis.indices.index(y_value)
+
+        if target is None:
+            value = trial.value
+        else:
+            value = target(trial)
+
+        if isinstance(value, int):
+            value = float(value)
+        elif not isinstance(value, float):
+            raise ValueError(
+                f"Trial{trial.number} has COMPLETE state, but its target value is non-numeric."
+            )
+        z_values[(x_i, y_i)] = value
+
+    return _SubContourInfo(xaxis=xaxis, yaxis=yaxis, z_values=z_values)
+
+
+def _get_axis_info(trials: List[FrozenTrial], param_name: str) -> _AxisInfo:
+    values: List[Union[str, float, None]]
+    if _is_numerical(trials, param_name):
+        values = [t.params.get(param_name) for t in trials]
+    else:
+        values = [
+            str(t.params.get(param_name)) if param_name in t.params else None for t in trials
+        ]
+
+    min_value = min([v for v in values if v is not None])
+    max_value = max([v for v in values if v is not None])
+
+    if _is_log_scale(trials, param_name):
+        min_value = float(min_value)
+        max_value = float(max_value)
+        padding = (math.log10(max_value) - math.log10(min_value)) * PADDING_RATIO
+        min_value = math.pow(10, math.log10(min_value) - padding)
+        max_value = math.pow(10, math.log10(max_value) + padding)
+        is_log = True
+        is_cat = False
+
+    elif _is_numerical(trials, param_name):
+        min_value = float(min_value)
+        max_value = float(max_value)
+        padding = (max_value - min_value) * PADDING_RATIO
+        min_value = min_value - padding
+        max_value = max_value + padding
+        is_log = False
+        is_cat = False
+
+    else:
+        span = len(set(values)) - 1
```

Comment:
In my understanding, when we add `None` as a dummy variable by https://github.com/optuna/optuna/pull/3682/files#diff-ebd7257ff163066174317cc1f78e73959751c319618bf9352ff5a31a464d0d5aR341, this `span` considers `None` value too. So we also need to ignore `None` here too to calculate `span`:
```suggestion
        unique_values = set(values)
        span = len(unique_values) - 1
        if None in unique_values:
            span -= 1
```

We can see the unnecessary padding by running the following code on this PR:
```python
import optuna


def objective(trial):
    x = trial.suggest_float("x", -100, 100)
    if trial.number <= 5:
        y = trial.suggest_categorical("y", [None, -1, 0, 1]) 
        if y is None:
            y = 1
    
    return x ** 2 + 1


sampler = optuna.samplers.TPESampler(seed=12)
study = optuna.create_study(sampler=sampler)
study.optimize(objective, n_trials=10)

optuna.visualization.plot_contour(study)
```

The code above generates the following plot:

![newplot](https://user-images.githubusercontent.com/7121753/175906112-77082317-272d-463a-b57e-c666df19819d.png)


Note that this bug is actually in the latest master branch too, so we can fix it by this PR.

Reply:
Thanks. I confirmed that the code snippet generates the following appropriate figure.
![cat](https://user-images.githubusercontent.com/38826298/176388626-20acb483-a3de-4684-8712-a0a15267f106.png)